File: backend/legalmind/domain/document_types.py
"""Document Type vocabulary — locked Step 6.

Step 6 fixes the initial V1 Document Types and distinguishes them from
Legal/Regulatory References ("These are not contract types"). A document is
additionally classified by source — ``Organization`` or ``Counterparty`` — and
Step 28's Requirement Model gives every Requirement a ``Document Type``, which is
what scopes a Requirement to the kind of paper it applies to. The worked example
in the locked comparison model is exactly this pairing: *"ABC MSA — Compared with
→ LeapSwitch Standard MSA."*

Why this is code and not a database enum: the locked physical schema (42.7) gives
the ``requirements`` table no document-type column and defines no
``document_types`` table, while Step 28 and Step 23 lock Document Type as a
concept and an admin-managed configuration area. That divergence is registered as
**C-13** and is not resolved here. Owner decision Q2 (2026-08-19) chose the
`D-3` route — the value lives in the Company Standard's ``configuration`` JSONB,
validated against this vocabulary in tested code, so no locked table changes.
A DB enum would be a schema change and would pre-empt C-13's resolution.
"""

from __future__ import annotations

# Locked Step 6's ten types, EXTENDED by the three Section 31 types (owner,
# 2026-09-18 — see below). Order matters only for error messages; membership is what
# is enforced.
#
# THE SECTION 31 EXTENSION.
#
# Legal Constitution L1.10 §31 "extends the Constitution's clause-wise structure to
# Partner Agreements, Vendor Agreements, Distribution Agreements, Purchase Orders,
# Amendments/Addenda, and Other Agreements", and tags each position with an
# "Applicable Document Types" line naming them. Locked Step 6 carried none of the
# first three, so no Company Standard could be typed to them and the importer refused
# every attempt: the Constitution stated positions the system had no vocabulary to
# hold.
#
# The visible consequence was a live defect (2026-09-18): "what is written about
# partner agreement in the constitution" was answered with three MSA standards,
# because no Partner Agreement position existed anywhere in the corpus. Registered as
# C-23 and resolved by the owner the same day — Constitution-final text is ratified,
# and every type the Constitution carries text for must answer from its own positions.
#
# Two of §31's six types needed nothing added:
#   * Purchase Order -> ORDER_FORM, already here and already how this repository
#     models it (the two proposed §31.11 standards are typed ORDER_FORM and coded
#     `PO-*`). A separate PURCHASE_ORDER type would have split one concept in two.
#   * Amendment/Addendum -> AMENDMENT, already here.
# "Other Agreements" (§31.13) adds no clauses of its own — it routes to the existing
# general sections — so it stays OTHER.
#
# EVIDENTIARY GRADE IS NOT CARRIED HERE. It lives in each standard's
# `configuration.constitution.basis`, whose vocabulary already distinguishes
# COMPANY_APPROVED from LEGALMIND_RULE (§31.9/§31.10's practice-based rules, "never
# Acceptable by guess") and NOT_ADOPTED (§31.6a, "never a current rule"). A document
# type says which paper a position governs; it never says how well evidenced that
# position is, and adding a type here asserts nothing about the text behind it.
DOCUMENT_TYPES: tuple[str, ...] = (
    "MSA",            # Master Services Agreement
    "NDA",            # Non-Disclosure Agreement
    "TOS",            # Terms of Service
    "SLA",            # Service Level Agreement
    "DPA",            # Data Processing Agreement
    "AUP",            # Acceptable Use Policy
    "PRIVACY_POLICY",
    "ORDER_FORM",     # also Constitution §31.11's Purchase Order (PO)
    "AMENDMENT",      # Amendment / Addendum — Constitution §31.12
    "PARTNER_AGREEMENT",       # Constitution §31.3-§31.8 (owner, 2026-09-18)
    "VENDOR_AGREEMENT",        # Constitution §31.9
    "DISTRIBUTION_AGREEMENT",  # Constitution §31.10
    "OTHER",
)

_DOCUMENT_TYPE_SET = frozenset(DOCUMENT_TYPES)


def readable(document_type: str) -> str:
    """The type as a reader sees it: `PARTNER_AGREEMENT` -> "Partner Agreement";
    `MSA` stays `MSA`.

    Needed the moment Step 6 gained a multi-word code (`AM-72`). A raw code composed
    into reader-facing text is not just ugly — `position_chunks` carries the type into
    the chunk body, and the Domain A egress screen reads any `WORD_WORD` token as an
    internal locator (an env var or a repo path), so `(PARTNER_AGREEMENT)` made the
    whole corpus refuse to egress. `ORDER_FORM` and `PRIVACY_POLICY` were the same
    landmine, unarmed only because no ratified standard was typed to either.

    Initialisms stay initialisms because that is what people call them.
    """
    return (document_type if document_type.isupper() and "_" not in document_type
            else document_type.replace("_", " ").title())


class UnknownDocumentType(ValueError):
    """A value outside locked Step 6's vocabulary.

    Raised rather than coerced: a wrong document type silently accepted would
    load the wrong baseline against a counterparty document, which is exactly
    the class of quiet error ENG-09 exists to prevent.
    """


def is_document_type(value: object) -> bool:
    return isinstance(value, str) and value in _DOCUMENT_TYPE_SET


def validate_document_type(value: object) -> str:
    """Return the value if it is a locked Step 6 type; raise otherwise.

    No normalisation is performed — ``"msa"`` is refused, not upcased. The
    vocabulary is a controlled legal classification, and a caller supplying a
    near-miss should be corrected at the boundary rather than guessed at.
    """
    if not is_document_type(value):
        raise UnknownDocumentType(
            f"unknown document type {value!r}; locked Step 6 defines exactly: "
            + ", ".join(DOCUMENT_TYPES))
    return value  # type: ignore[return-value]  # narrowed by is_document_type


# Document SOURCE — the second axis locked Step 6 names beside Type: "A document
# can be classified by source: Organization / Counterparty" (all_lock.md:535).
# "Can", not "must": the axis is optional, and DOC-06 leaves it "untouched", so
# this is the vocabulary only. Declared by the uploader, never inferred (the same
# reasoning as Q9 for Type), and stored in `document_versions.metadata` — the
# locked 42.4 JSONB — because a document's source can differ between versions:
# v1 our template, v2 the counterparty's redline.
DOCUMENT_SOURCES: tuple[str, ...] = ("ORGANIZATION", "COUNTERPARTY")

_DOCUMENT_SOURCE_SET = frozenset(DOCUMENT_SOURCES)


def is_document_source(value: object) -> bool:
    return isinstance(value, str) and value in _DOCUMENT_SOURCE_SET

File: frontend/src/lib/documentTypes.ts
/**
 * Locked Step 6's Document Type vocabulary — presentation copy of
 * `backend/legalmind/domain/document_types.py`, which is the source of truth
 * and validates every value server-side. `backend/tests/test_frontend_vocabulary.py`
 * asserts the two lists are identical, so a drift on either side fails CI.
 *
 * Owner ruling Q9 (2026-08-19): the type is DECLARED by the uploader, never
 * inferred — analysis refuses an undeclared document rather than guessing which
 * baseline to compare it against. The intake screen therefore makes this the one
 * prominent required choice.
 */
export const DOCUMENT_TYPES: ReadonlyArray<{ code: string; label: string }> = [
  { code: "MSA", label: "Master Services Agreement" },
  { code: "NDA", label: "Non-Disclosure Agreement" },
  { code: "TOS", label: "Terms of Service" },
  { code: "SLA", label: "Service Level Agreement" },
  { code: "DPA", label: "Data Processing Agreement" },
  { code: "AUP", label: "Acceptable Use Policy" },
  { code: "PRIVACY_POLICY", label: "Privacy Policy" },
  { code: "ORDER_FORM", label: "Order Form" },
  { code: "AMENDMENT", label: "Amendment / Addendum" },
  { code: "OTHER", label: "Other" },
];

export function documentTypeLabel(code: string | null | undefined): string {
  return DOCUMENT_TYPES.find((t) => t.code === code)?.label ?? code ?? "Type not declared";
}

/**
 * A name for the contract, derived from the file the user chose — an editable
 * DEFAULT, never a demand (2026-08-31 UX correction): the filename already
 * carries the natural label, and identity/audit live on ids and the preserved
 * original_filename, not on this display name.
 */
export function nameFromFilename(filename: string): string {
  const stem = filename.replace(/\.[A-Za-z0-9]{1,5}$/, "");
  const tidied = stem.replace(/[_\-.]+/g, " ").replace(/\s+/g, " ").trim();
  return tidied || filename;
}

/** Filename tokens that plainly suggest a Step 6 type. PRESENTATION ONLY: the
 *  hint is shown beside the (empty) select and applied only by the user's own
 *  click — the declaration stays a human act (owner Q9), nothing is inferred
 *  into the record. */
const TYPE_TOKENS: ReadonlyArray<[string, string]> = [
  ["msa", "MSA"], ["nda", "NDA"], ["tos", "TOS"], ["sla", "SLA"],
  ["dpa", "DPA"], ["aup", "AUP"], ["privacy", "PRIVACY_POLICY"],
  ["order", "ORDER_FORM"], ["amendment", "AMENDMENT"], ["addendum", "AMENDMENT"],
];

export function typeHintFromFilename(filename: string): string | null {
  const tokens = filename.toLowerCase().split(/[^a-z0-9]+/);
  for (const [token, code] of TYPE_TOKENS) {
    if (tokens.includes(token)) return code;
  }
  return null;
}

/**
 * `§` in front of a section number — once, never twice.
 *
 * `Evidence.section_number` carries the document's own numbering, and real
 * documents supply it both ways: "17.2" from one parser, "§17.2" from another.
 * Prepending unconditionally rendered "§§1" on every clause of an MSA whose
 * headings already carry the sign.
 *
 * It lives here, shared, because the first fix patched two of the four callers
 * and the other two kept printing "§§" — a rule copied into every caller is a
 * rule that gets fixed in some of them.
 */
export function sectionRef(sectionNumber: string | null | undefined): string | null {
  // The number exactly as the document states it (owner, 2026-09-10: "DO NOT
  // add the § symbol… the document itself is the source of truth"). A sign the
  // document carries stays; none is ever added.
  const trimmed = (sectionNumber ?? "").trim();
  return trimmed || null;
}

/**
 * Step 6's second axis — "A document **can** be classified by source:
 * Organization / Counterparty" (all_lock.md:535). Optional, and declared by the
 * uploader, never inferred (the same reasoning as Q9 for Type). Presentation
 * copy of `DOCUMENT_SOURCES` in the backend's `document_types.py`;
 * `test_frontend_vocabulary.py` asserts the two agree. Keyed `value`, not
 * `code`, so the Step 6 TYPE sync regex never picks these up by mistake.
 */
export const DOCUMENT_SOURCES: ReadonlyArray<{ value: string; label: string }> = [
  { value: "ORGANIZATION", label: "Organization — our document" },
  { value: "COUNTERPARTY", label: "Counterparty — their document" },
];

export function documentSourceLabel(value: string | null | undefined): string {
  return DOCUMENT_SOURCES.find((s) => s.value === value)?.label ?? value ?? "Source not declared";
}

/**
 * The source axis in a reviewer's words, for a chip (2026-09-06). The manager's
 * question is "whose paper is this?" — "Organization"/"Counterparty" answers it
 * in the vocabulary Step 6 uses, but not in the words a reviewer thinks in.
 */
export function documentSourceChip(value: string | null | undefined): string | null {
  if (value === "ORGANIZATION") return "Our document";
  if (value === "COUNTERPARTY") return "Their document";
  return null;
}
